### DISCUSSION TITLE
persistent caching

### DISCUSSION THREAD

**Discussion by hajes** [ranaroussi/yfinance#2148] [Q&A]
hi guys,
I am using yfinance for years in my investing career. Option screener, stock valuations. I started to experiment with caching as suggested in documentation. It is great for testing. It seems to store all data fetched from Yahoo...well at least for an hour or so.

In previous attempts, I simply stored all in pandas/sqlite3. Update only what is needed without fetching all data again.

I came across `yfinance-cache` clone. It doesn't work with historical data. It makes me wonder one can setup longer period of caching. Let's say 1 month of data cached. I run the script once a week, and wish to avoid fetching all over again.

Any suggestions, please?

**Comment by Chenghao999** [comment]
Worth setting expectations first: yfinance's persistent cache does **not** cache price data. In 1.7.0 `yfinance/cache.py` is a peewee/SQLite store for three pieces of *metadata* only — timezone strings (`_TzCache`), Yahoo cookies (`_CookieCache`), and the ISIN↔ticker map (`_ISINCache`). No OHLCV. So there's no switch to make it hold a month of history; `yf.set_cache_location()` / `set_tz_cache_location()` only move where that metadata DB lives.

For a weekly job that should fetch only the new tail, wrap it yourself:

```python
import pandas as pd, yfinance as yf
from pathlib import Path
from datetime import datetime, timedelta

CACHE = Path("~/.cache/px").expanduser(); CACHE.mkdir(parents=True, exist_ok=True)

def history(ticker, max_age=timedelta(days=7)):
    f = CACHE / f"{ticker}.parquet"
    if not f.exists():
        df = yf.Ticker(ticker).history(period="max")
        df.to_parquet(f); return df
    df = pd.read_parquet(f)
    if datetime.now(df.index.tz) - df.index[-1] > max_age:
        fresh = yf.Ticker(ticker).history(start=df.index[-1] + timedelta(days=1))
        if not fresh.empty:
            df = pd.concat([df, fresh[~fresh.index.isin(df.index)]]).sort_index()
            df.to_parquet(f)
    return df
```

Two things that bite in practice:

- **Adjustment drift.** `auto_adjust=True` is the default (`yf.download`'s signature shows it; `Ticker.history` behaves the same). Adjusted prices are recomputed retroactively on every split/dividend, so appending only the new tail leaves two different bases in one frame — quietly wrong for valuation work. Either re-pull the whole ticker when a corporate action lands, or store unadjusted prices plus the adjustment factors and compute adjusted values yourself.
- **Freshness is yours to define.** yfinance has no notion of "cache for N days"; 7 days just matches your weekly rhythm.

If fewer requests is the real goal, batch with `yf.download(tickers=[...])` instead of looping tickers — and keep the index tz consistent across parquet writes, or you'll hit dtype trouble later.

**Comment by ValueRaider** [comment]
> I came across yfinance-cache clone. It doesn't work with historical data.

Report a bug. It's supposed to work.
